### ISSUE TITLE
dask.Bag.groupby with hash seed in a distributed environment

### ISSUE THREAD

**Issue by chmp** [dask/dask#1850]
Hi, 

recently I played around with `dask.Bag.groupby`. However, with python 3.5 I quickly ran into problems, since the hash seeds are [randomly set][hash-seed]. In particular for strings the hashes differ between workers even for identical values. Therefore, `.groupby` will give wrong results. This problem does not occur for integer values, since their hashes do stay the same for different seeds.

While I am aware that setting the `PYTHONHASHSEED` environment variable should fix the problem, I am wondering whether it would be a good idea to implement a custom hash function that is guaranteed to give the same results between workers independent of environment variables.

A simple example to illustrate the problem is:

```python
# used versions:
# python: 3.5.1
# distributed version: 1.14.3
# dask version: 0.12.0

from distributed import Client
from dask.bag import from_sequence
from toolz import identity

# with 4 connected workers
client = Client('127.0.0.1:8786')

print(
    'hashes', 
    set(from_sequence(['foo'] * 200, npartitions=20).map(hash).compute(get=client.get)),
)
# example output:
# hashes {-3084888748298431351, -75493216565243888, 5604312191287608001, -2400793391545145946}

print(
    'len', 
    len(from_sequence(['foo'] * 200, npartitions=20).groupby(identity).compute(get=client.get))
)

# example output:
# len 4 
```

Regards,

Christopher

[hash-seed]: http://stackoverflow.com/a/27522708

**Comment by mrocklin** [comment]
Ah, good catch.  I'll admit that I'm tempted by setting the hash seed, mostly because it will be robust across all python types that implement `__hash__`.  Open to other possibilities though.

**Comment by chmp** [comment]
I tend to agree with you, building a custom hash system sounds like a nightmare. Then, documenting the issue in the distributed docs should be enough. If I find time, I'll open a corresponding PR tomorrow.

**Comment by mrocklin** [comment]
Is it possible to set the hash seed from within the process?  It would be nice to have a sane default.

**Comment by pitrou** [comment]
No, it's not possible. The hash seed is set at startup and can't be changed afterwards, because hash values are cached in a large number of diverse places.

**Comment by mrocklin** [comment]
Can you think of any clever solutions here?  In common practice the user does start a new Python process with `dask-worker ...`.  Is there a more pleasant solution than the following:

    export PYTHONHASHSEED=1234; dask-worker ...

**Comment by pitrou** [comment]
Well, an obvious solution would be to turn `dask-worker` into a shim that spaws the actual Python process. But that's not very pretty and more fragile to maintain.
Is there a way to avoid passing hash values between processes?

**Comment by mrocklin** [comment]
We could stop using the Python `hash` function within operations like `dask.bag.Bag.groupby`, replacing it with a consistent hash.  This has the obvious drawback of not working on arbitrary Python objects.

My guess is that this isn't doable, but can we set the hash-seed for the forkserver that we use to spawn workers from the nanny?

**Comment by mrocklin** [comment]
Er, should read "that we use to fork workers from the nanny".  Improper use of the word spawn.

**Comment by pitrou** [comment]
Theoretically the forkserver could allow that, but there's no API for it, unfortunately.

As for the hash, I was wondering if it's possible to still use the Python hash function without communication hash values between processes. I find it surprising that you're communicating hash values, rather than the results of e.g. the grouper function. After all, hash values can suffer collisions, so they're not enough in themselves.

**Comment by mrocklin** [comment]
If two values are placed into the same group by the hashing function (mod some desired number of groups) then we need to make sure that those two values end up in the same group on the other side.  Additionally we eventually need to combine these groups:

```python
In [1]: from string import ascii_lowercase
In [2]: from toolz import groupby
In [3]: groupby(lambda x: hash(x) % 4, ascii_lowercase)
Out[3]: 
{0: ['a', 'c', 'd', 'e', 'f', 'g', 's', 'v', 'z'],
 1: ['j', 'k', 'u', 'x'],
 2: ['b', 'h', 'o', 'p', 'q', 'r', 't', 'w'],
 3: ['i', 'l', 'm', 'n', 'y']}
```

```python
In [1]: from toolz import groupby
In [2]: from string import ascii_lowercase
In [3]: groupby(lambda x: hash(x) % 4, ascii_lowercase)
Out[3]: 
{0: ['b', 'd', 'g', 'j', 'p', 's', 'x'],
 1: ['c', 'k', 'l', 'o', 'q', 't', 'w', 'y'],
 2: ['h', 'm', 'z'],
 3: ['a', 'e', 'f', 'i', 'n', 'r', 'u', 'v']}
```

**Comment by pitrou** [comment]
I guess I'll have to take a look at groupby's internals to understand a little more what it's doing.

**Comment by chmp** [comment]
Just to confirm: when setting the hash seed before starting the worker, everything works as expected: there is only a single hash value and the groupby result contains a single item.

**Comment by mrocklin** [comment]
> I guess I'll have to take a look at groupby's internals to understand a little more what it's doing.

Groupby is roughly the following:

```python
def groupby(func, sequence):
    d = defaultdict(list)
    for item in sequence:
        key = func(item)
        d[key].append(item)
    return d
```

**Comment by pitrou** [comment]
Sure, but that doesn't answer my question why the hash values (rather than grouping keys) have to be communicated between processes ;-)

**Comment by mrocklin** [comment]
I may not understand.  In this case the hash value mod n is the grouping key.

**Comment by pitrou** [comment]
Hmm. Based on the example posted initially, this doesn't seem to be the case, or perhaps I'm misunderstanding:
```
print(
    'len', 
    len(from_sequence(['foo'] * 200, npartitions=20).groupby(identity).compute(get=client.get))
)

# example output:
# len 4
```

**Comment by chmp** [comment]
From my understanding the distributed groupby works by first sorting the items into partitions and then perform a `toolz.groupby` within each partition separately. The hash value is used to map the key to the partition index as `hash(key) % number_of_input_partitions`. If the hash value were unique, it would be correct to then group independently in each partition. In my example, each worker computes a different hash value even for identical inputs and therefore there is a high chance that the values are scattered over as many partitions as there are workers. 

However, I have to admit that I find the `groupby_tasks` implementation rather hard to follow and am note 100 % sure I understand it correctly.

**Comment by mrocklin** [comment]
Ah, yes, I see your point.  After we create groups per partition we now need to communicate those intermediate groups together into different output partitions.  For example if they keys are numbers and we want to produce two output partitions then we move all even-keyed groups to one partition and all odd-keyed groups to another.  We want to use a consistent hash function to evenly and predictably coordinate which groups should end up in which output partition.

**Comment by mrocklin** [comment]
groupby_tasks is indeed fairly complex.  It is difficult to move data from a possibly large number of input partitions to a possibly large number of output partitions without generating N^2 tasks.  @pitrou , this code looks a bit like the array rechunking problem you optimized earlier.

**Comment by mrocklin** [comment]
That's concerning.  Single-machine shuffles write to a predefined location on disk.  This location isn't being cleared after the final read.  We should probably make writes idempotent, perhaps by passing along a unique identifier (though there are certainly other solutions).

@pitrou can I ask you to make a separate issue for this?

**Comment by pitrou** [comment]
Well it's already in a separate issue :-) https://github.com/dask/dask/issues/1851

**Comment by mrocklin** [comment]
The single-machine shuffle issue was resolved.

**Comment by jakirkham** [comment]
Is this still an issue?

**Comment by chmp** [comment]
To be honest, I didn't use dask for a while now and have not kept up with the internals. The example I posted above gives the expected result when run with current package versions. From my side, I would close this issue.

**Comment by jakirkham** [comment]
Great! Thanks for checking. 🙂
